main.go

```diff
@@ -14,7 +15,386 @@
+					tlsconfig.MTLSClientConfig(source, source, tlsconfig.AuthorizeAny()),
+				),
+			),
+		),
+	)
+
+	if config.RegisterService {
+		for _, serviceName := range config.ServiceNames {
+			nsRegistryClient := registryclient.NewNetworkServiceRegistryClient(ctx, &config.ConnectTo, registryclient.WithDialOptions(clientOptions...))
+			_, err = nsRegistryClient.Register(ctx, &registryapi.NetworkService{
+				Name:    serviceName,
+				Payload: config.Payload,
+			})
+
+			if err != nil {
+				log.FromContext(ctx).Fatalf("unable to register ns %+v", err)
+			}
+		}
+	}
+
+	nseRegistryClient := registryclient.NewNetworkServiceEndpointRegistryClient(
+		ctx,
+		&config.ConnectTo,
+		registryclient.WithDialOptions(clientOptions...),
+		registryclient.WithNSEAdditionalFunctionality(
+			registrysendfd.NewNetworkServiceEndpointRegistryClient(),
+		),
+	)
+	nse := getNseEndpoint(config, listenOn)
+	nse, err = nseRegistryClient.Register(ctx, nse)
+	logrus.Infof("nse: %+v", nse)
+
+	if err != nil {
+		log.FromContext(ctx).Fatalf("unable to register nse %+v", err)
+	}
+
+	// ********************************************************************************
+	log.FromContext(ctx).Infof("startup completed in %v", time.Since(starttime))
+	// ********************************************************************************
+
+	// wait for server to exit
+	<-ctx.Done()
+}
+
+func getNseEndpoint(config *Config, listenOn fmt.Stringer) *registryapi.NetworkServiceEndpoint {
+	nse := &registryapi.NetworkServiceEndpoint{
+		Name:                 config.Name,
+		NetworkServiceNames:  config.ServiceNames,
+		NetworkServiceLabels: make(map[string]*registryapi.NetworkServiceLabels),
+		Url:                  listenOn.String(),
+	}
+	for _, serviceName := range config.ServiceNames {
+		nse.NetworkServiceLabels[serviceName] = &registryapi.NetworkServiceLabels{Labels: config.Labels}
+	}
+	return nse
+}
+
+func getSriovTokenServerChainElement(ctx context.Context) (tokenServer networkservice.NetworkServiceServer) {
+	sriovTokens := tokens.FromEnv(os.Environ())
+	switch len(sriovTokens) {
+	case 0:
+		tokenServer = null.NewServer()
+	case 1:
+		var tokenKey string
+		for tokenKey = range sriovTokens {
+			break
+		}
+		tokenServer = token.NewServer(tokenKey)
+	default:
+		log.FromContext(ctx).Fatalf("endpoint must be configured with none or only one sriov resource")
+	}
+	return
+}
+
+func exitOnErr(ctx context.Context, cancel context.CancelFunc, errCh <-chan error) {
+	// If we already have an error, log it and exit
+	select {
+	case err := <-errCh:
+		log.FromContext(ctx).Fatal(err)
+	default:
+	}
+	// Otherwise wait for an error in the background to log and cancel
+	go func(ctx context.Context, errCh <-chan error) {
+		err := <-errCh
+		log.FromContext(ctx).Error(err)
+		cancel()
+	}(ctx, errCh)
+}
+
+// newPolicyRoutesGetter - watches the config file and dynamically updates policy routes.
+func newPolicyRoutesGetter(ctx context.Context, configPath string) *policyRoutesGetter {
```

Comment:
This whole function is probably unneeded in this NSE.

Reply:
Done.